### PR TITLE
Adds cache mechanism to middleware

### PR DESCRIPTION
**Pull request by Willianvdv**
This PR adds a mechanism to add caching that will store middleware results. In our project, I'm using it like so:
``` 
Arel.middleware.apply([middleware_one], cache: Rails.cache) do
  User.all.load
end
```

### CODE REVIEW

File: lib/arel/middleware/chain.rb [hunk 1]
```diff
@@ -21,9 +21,9 @@ def execute(sql, binds = [], &execute_sql)
         check_middleware_recursion(sql)
 
         updated_context = context.merge(original_sql: sql)
-        enhanced_arel = Arel.enhance(Arel.sql_to_arel(sql, binds: binds))
+        arel = Arel.sql_to_arel(sql, binds: binds)
```
Comment (Willianvdv) [inline lib/arel/middleware/chain.rb]: As I like to use the SQL (and binds) as the cache key, making this sql_to_arel part of the executor and make it part of the cached value will save some unnecessary operations. 

However, I'm not sure how to do this without changing the DSL of the middleware.

Comment (Willianvdv) [reply lib/arel/middleware/chain.rb]: Now that I'm thinking about it, maybe having a middleware that does the caching isn't the easiest (mostly implementation wise) solution. Instead, I could add some first-class caching support to the chain that just wraps all this execution stuff.

Comment (Willianvdv) [reply lib/arel/middleware/chain.rb]: and that's what I did

File: lib/arel/middleware/database_executor.rb [hunk 2]
```diff
@@ -46,7 +48,12 @@ def call(next_arel)
 
       def execute_sql(next_arel)
         sql, binds = next_arel.to_sql_and_binds
+
+        # TODO: don't set cache if the binds != context[:original_binds]
```
Comment (Willianvdv) [inline lib/arel/middleware/database_executor.rb]: I have to verify but I think we should not cache if the bind params are reordered.

File: lib/arel/middleware/chain.rb (hunk 1 above)
Comment (mvgijssel) [reply lib/arel/middleware/chain.rb]: Awesome!

File: lib/arel/middleware/database_executor.rb [hunk 3]
```diff
@@ -46,7 +46,16 @@ def call(next_arel)
 
       def execute_sql(next_arel)
         sql, binds = next_arel.to_sql_and_binds
+
+        context[:cache].set(
```
Comment (Willianvdv) [inline lib/arel/middleware/database_executor.rb]: As a caching instance can be shared by multiple chains, I guess the transformations done on this query should be part of the cache key.

Comment (Willianvdv) [reply lib/arel/middleware/database_executor.rb]: On a different note, I'm not too excited about passing lots of responsibility and internal transformation details to the caching implementation. I'm considering adding a bit more logic to the chain (this module) to make it more opinionated about what to cache.

File: lib/arel/middleware/chain.rb [hunk 4]
```diff
@@ -109,8 +115,23 @@ def to_sql(type, &block)
 
       private
 
-      def continue_chain(middleware, context, &block)
-        new_chain = Arel::Middleware::Chain.new(middleware, context)
+      def execute_with_middleware(sql, binds, execute_sql)
```
Comment (Willianvdv) [inline lib/arel/middleware/chain.rb:119]: The linter made me extract this bit

File: lib/arel/middleware/chain.rb [hunk 5]
```diff
@@ -1,31 +1,65 @@
...
+
+      def set(transformed_sql:, transformed_binds:, original_sql:, original_binds:)
+        # The order of binds was changed and therefore we can't reuse the query
+        return if transformed_binds != original_binds
+
+        cache.set(cache_key(original_sql), transformed_sql)
+      end
+
+      def cache_key(original_sql)
+        # TODO: Add the active middleware to the cache key
```
Comment (Willianvdv) [inline lib/arel/middleware/chain.rb]: @mvgijssel do you know if it's possible to get the whole middleware chain? The thing I like to achieve is that the cache key becomes something like:
```
SomeMiddleware::AnotherMiddleware::SELECT * FROM kerks;
```

**Comment by mvgijssel** [comment]
Thinking about this, do we actually need the bind values to be a part of the cache key? Can't we only the input SQL?

**Comment by Willianvdv** [comment]
> Thinking about this, do we actually need the bind values to be a part of the cache key? Can't we only the input SQL?

Yes, the query is cached using the input query as the cache key. Caching only happens if the order of the bind params didn't change.

File: lib/arel/middleware/chain.rb (hunk 5 above)
Comment (mvgijssel) [reply lib/arel/middleware/chain.rb]: Something like this?

```ruby
class Foo; def call(arel, next_middleware); next_middleware(arel); end; end
class Bar; def call(arel, next_middleware); next_middleware(arel); end; end

Arel.middleware.apply([Foo, Bar]) { Arel.middleware.current.map(&:name) }
```

Comment (Willianvdv) [reply lib/arel/middleware/chain.rb]: Let me try!

File: lib/arel/middleware/chain.rb [hunk 6]
```diff
@@ -1,31 +1,73 @@
 module Arel
   module Middleware
+    module NoOpCache
```
Comment (mvgijssel) [inline lib/arel/middleware/chain.rb]: Can you move this to a separate file?

File: lib/arel/middleware/chain.rb [hunk 7]
```diff
@@ -1,31 +1,73 @@
...
+      def self.get(key); end
+
+      def self.set(key, sql); end
+    end
+
+    class CacheAccessor
```
Comment (mvgijssel) [inline lib/arel/middleware/chain.rb]: Same

**Review by mvgijssel — APPROVED** [review]

File: .github/workflows/develop.yml [hunk 8]
```diff
@@ -47,8 +47,10 @@ jobs:
     - name: Rebase
       if: matrix.rebase
       run: |
-        git config --global user.email "ci@github.com"
-        git config --global user.name "GitHub CI"
+        export GIT_AUTHOR_NAME=${CI_COMMIT_AUTHOR_NAME=GithubCI}
```
Comment (Willianvdv) [inline .github/workflows/develop.yml:50]: Without this change, the ci failed because the email and name were not set.

**Pull request merged**